### PR TITLE
[SPARK-48754] Separate session code from sql module

### PR DESCRIPTION
**Pull request by mathiasschw-db**
### What changes were proposed in this pull request?
This pulls the session code into a separate package, enforcing the layered structure that is defined in the README. This is the change that we prepared by introducing the sparkExecutor interface earlier

### Why are the changes needed?
This improves the package structure by making it reflect the dependencies described in the README directly

### Does this PR introduce _any_ user-facing change?
It moves the session code to a new package but introduces no behavioral changes

### How was this patch tested?
Ran existing unit tests

### CODE REVIEW

File: client/mocks/mocks.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package mocks
+
+import (
+	"context"
+
+	proto "github.com/apache/spark-connect-go/v3.5/internal/generated"
```
Comment (grundprinzip) [inline client/mocks/mocks.go:6]:
```suggestion
	proto "github.com/apache/spark-connect-go/v35/internal/generated"
```

**Comment by grundprinzip** [comment]
I think you need to pull from master again to fix the merge conflicts.

**Comment by mathiasschw-db** [comment]
Yes, looks like my change was overtaken by a different one

**Comment by mathiasschw-db** [comment]
Rebased

**Review by grundprinzip — COMMENTED** [review]
So I have a question here, in Spark/PySpark the `SparkSession` is part of the SQL module, but the client lives outside. 

Is there a benefit of moving this to a different package (just from the least surprise principle)

**Comment by mathiasschw-db** [comment]
So basically, I am trying to separate some very tightly coupled components here. You know more about how folks expect to use this so we can keep it in on package if that is preferable.

I am trying to make the codebase (to some degree) reflect the dependency diagram from README.md but increasingly decoupling the very highly coupled components of the package. The current dependencies look like this:
```
    +-------------------+                                                                              
    |                   |                                                                              
    |  sparkSessionImpl |                                                                              
    |                   |                                                                              
    +-------------------+                                                                              
              |                                                                                        
              |                                                                                        
              +                                                                                        
    +-------------------+                                                                              
    |                   |                                                                              
    | dataFrameImpl     |                                                                              
    |                   |                                                                              
    +-------------------+                                                                              
              |                                                                                        
              |                                                                                        
              +                                                                                        
+---------------------------+               +----------------+                                         
|                           |               |                |                                         
| SparkConnectServiceClient |--------------+|  Spark Driver  |                                         
|                           |               |                |                                         
+---------------------------+               +----------------+
```
That doesn't quite match what the README.md file says (session and dataframe are reversed).

**Review by grundprinzip — APPROVED** [review]
I think the project layout is still going to be quite fluid, and we're going to shuffle things around a bit until we have something that provides the ergonomics that we're looking for. 

Since we don't have a release yet, we can do that.

I will create a follow-up PR that will rename some of the packages to follow the Spark convetions a bit more.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by mathiasschw-db** [comment]
Sounds good. Yes, I agree that this isn't final. It is easier to shuffle around in smaller pieces even if we end up merging it again. In the mean time, it enforces som discipline around avoiding the circular dependencies that I broke up

**Pull request closed without merging**